design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,119 @@
+will be saved in a designated path used by the GC client to list branch's uncommitted objects.  
+For the purpose of this document we'll call this the `GC commit` (feel free to suggest a better term :) )
+
+#### Reading data from lakeFS
+
+Reading data from lakeFS will be similar to the current GC process - using the SSTable reader to quickly list branch objects.
+The above-mentioned API will enable doing so for all branch objects (committed and uncommitted).
+
+### GC Flows
+
+The following describe the GC process run flows in the branch scope:
+
+#### Flow 1: Clean Run
+
+1. Run PrepareUncommittedForGC
+2. Read all addresses from branch commits -> `lakeFS DF`
+3. Read all addresses from branch `GC commit` -> `lakeFS DF`
+4. Read all objects on branch path directly from object store (can be done in parallel by 'partition') -> `Branch DF`
+5. Subtract results `lakeFS DF` from `Branch DF`
+6. Filter files newer than < TOKEN_EXPIRY_TIME > and special paths
+7. The remainder is a list of files which can be safely removed
+8. Finally, save the current run's `GC commit` the last read partition and newest commit id in a designated location on the branch path
+
+#### Flow 2: Optimized Run
+
+Optimized run uses the previous GC run output, to perform a partial scan of the branch to remove uncommitted garbage.
+
+##### Step 1. Analyze Data and Perform Cleanup for old entries (GC client)
+
+1. Run PrepareUncommittedForGC
+2. Read addresses from branch's new commits (all new commits down to the last GC run commit id) -> `lakeFS DF`
+3. Read addresses from branch `GC commit` -> `lakeFS DF`
+4. Subtract results `lakeFS DF` from previous run's `GC commit`
+5. The result is a list of files that can be safely removed
+
+>**Note:** This step handles cases of objects that were uncommitted during previous GC run and are now deleted
+
+##### Step 2. Analyze Data and Perform Cleanup for new entries (GC client)
+1. Read all objects on branch path down to the previous run's last read partition (can be done in parallel by 'partition') -> `Branch DF`
+2. Subtract `lakeFS DF` from `Branch DF`
+3. Filter files newer than < TOKEN_EXPIRY_TIME > and special paths
+4. The remainder is a list of files which can be safely removed
+5. Finally, save the current run's `GC commit` the last read partition and newest commit id in a designated location on the branch path
+
+## Limitations
+
+* Since this solution relies on the new repository structure, it is not backwards compatible. Therefore, another solution will be required for existing 
+repositories
+* This solution partitions the branch data into segments that allow parallel listing of branch objects. In cases where
+the branch path is extremely large, it might still take a lot of time to perform a clean run.
+* For GC to work optimally, it must be executed on a timely basis. Long periods between runs might result in excessively long runs.
```

Comment:
Can we improve it by limiting GC cycles (time limit or number of objects to handle)?

Reply:
This will be a bit difficult IMO. Since object creation date is loosely coupled with commit date it might be very complicated to achieve

Reply:
This is an ops issue.  I might _prefer_ occasional long runs, or common short runs.

Reply:
Rephrased